### PR TITLE
Fixes #40 Implement LayoutBuilder subclass

### LINKED ISSUE #40: Implement LayoutBuilder Subclass
## Motivation
During planning of our MVP for usage of Layout Builder, we have identified a few workflow customizations we would like to implement. It would be desirable if we could gather these customizations to LayoutBuilder workflows inside one of our custom modules related to Layout.

## Proposed Resolution
It's recently become possible in Drupal 8.8.0 to override the LayoutBuilder class in use, and it might be useful to implement a LayoutBuilder subclass that can be used to hold these customizations, to go along with the new changes in #38 

Some initial targets for this might consist as follows:
- Responsive Sidebar/Modal sidebar (to make working with large block forms feasible)
- Remove the "Add a custom block" secondary step and proceed directly to choosing the block type.
- Ability to add needed layout css/js.

### PR DESCRIPTION
**Pull request by tadean**
This pull request builds on #39 and adds an child class for LayoutBuilder that attempts to provide a place to house customizations we have in mind. The LayoutBuilder subclass is swapped in as the active LayoutBuilder. Functionality thus far:

**Triggered library loading of js/css files independently in several contexts:**
- Off-canvas menu
- Layout Builder form
- Flexible Pages

**Some initial customizations:**
- Hide the floating "You are editing..." message.
- Move submit buttons to the bottom.
- Hide the help box (until we decide on our own help structure)
- Adding blocks goes directly to the block selection
- Widens the block configuration sidebar

## Related Issue

#40 #38 

## How Has This Been Tested?

In need of testing, and addition of new unit tests to account for the change.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tadean** [comment]
Todo:

- [ ] Style adjustments to account for conflicts with the Drupal off-screen css. These were working before when @akslay developed them, but being added as theme css rather than injected css changed the order of a few things.
- [ ] Ensure new layout builder only loads for flexible pages
- [ ] Further customizations?

File: modules/custom/az_blocks/az_blocks.libraries.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  dependencies:
+    - az_barrio/global-styling
+    - az_barrio/arizona-bootstrap
+az_blocks.layout:
+  version: VERSION
+  css:
+    theme:
+      css/az_blocks_layout.css: {}
+  dependencies:
+    - layout_builder/drupal.layout_builder
```
[the remark this answers is not in the document]
Comment (tadean) [reply modules/custom/az_blocks/az_blocks.libraries.yml]: Agreed, sounds good to me! Thanks @joeparsons !

[the remark this answers is not in the document]
Comment (tadean) [reply modules/custom/az_blocks/az_blocks.libraries.yml]: Regarding the individual block types, maybe we could do something like:
- Check each type of custom block that exists.
- For each custom block type, check if a css/js file with the same machine name exists in the module.
- If so, `hook_library_info_alter` it into the library definition?

[the remark this answers is not in the document]
Comment (joeparsons) [reply modules/custom/az_blocks/az_blocks.libraries.yml]: Sounds reasonable to me.

**Comment by joeparsons** [comment]
@tadean I noticed that the Layout Builder Styles module modifies some of the default Layout Builder provided routes to swap in their own forms.  Do we need to verify that the changes here aren't interfering with those?  (See https://git.drupalcode.org/project/layout_builder_styles/blob/8.x-1.x/src/EventSubscriber/RouteSubscriber.php)

**Comment by tadean** [comment]
Thanks! I think I mostly tried to create brand new module-specific routes to new destinations rather than altering existing ones, but I'll check it out.

Actually that method they're using there to override actually seems way more elegant than what I was doing; maybe there's some refactoring I can do to do something similar...

**Comment by tadean** [comment]
Merged in the new changes from master and resolved the `az_media` namespace merge conflict in configuration files, this branch had the original names for the media entities.

**Comment by joeparsons** [comment]
Seems it's possible to see a two-dot version of the diff between this branch and master here:
https://github.com/az-digital/az_quickstart/compare/master..feature/40

**Comment by tadean** [comment]
Merged in the new changes from the master branch. @joeparsons fix in #49 has cleared up the display suite deprecated code issue in `drupal-check`.

**Comment by tadean** [comment]
In some testing I found that block drag and drop wasn't working correctly with the ds layout plugins. I'll investigate. The blocks seem to be rendering outside the section region where drag and drop normally. Perhaps a template or theme issue.

- [ ] Drag and drop markup issue

**Comment by tadean** [comment]
It appears that the reason for drag and drop not working is that the display suite layout plugins are not rendering some necessary internal markup that layout builder is expecting.

File: modules/custom/az_blocks/az_blocks.libraries.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tadean) [reply modules/custom/az_blocks/az_blocks.libraries.yml]: In the latest commits, I've moved the `az_blocks` libraries over to the `az_flexible_page` module (with a rename.)

**Comment by tadean** [comment]
I've added code to `az_blocks` to support selectively adding optional block-level libraries that may or may not exist, in `hook_preprocess_block` with the following strategy:

- Check if the block is an inline content block
- If so, check what bundle the block is
- Check if `az_blocks` implements a matching library for that bundle, eg. `az_accordion` checking for a `az_blocks/az_blocks.az_accordion` library`.
- If found, attach that library to the block.

This is an attempt to set it up so that blocks can attempt to load their individual libraries on an as-needed basis. There are a few block-specific css rules in the `az_flexible_page.css` that still need extracting out into individual libraries to make use of this.

**Comment by tadean** [comment]
- [ ] Stub out existing block-specific CSS into new libraries to make use of the selective loading

**Comment by tadean** [comment]
Added a reusable `az_layouts.customize_inline_block_form` to `az_layouts` that allows for customizing the layout builder inline block forms. Currently the alteration does the following:

- Add an automatic title (following the format `Inline BLOCK_TYPE`, e.g. a `az_accordion` will have the hidden title `Inline Accordion`)
- Hides the layout builder Title field (Not any title field on the block itself)
- Unchecks the Display Title checkbox
- Hides the Display Title checkbox

This allows for flexible page blocks to be added with a title only being necessary if the block has its own title field, without a duplicate title form field:

image

**Comment by tadean** [comment]
Swapped to a more extensible method of choosing which layout offscreen canvases need to be wider, on a per-route basis.

**Comment by tadean** [comment]
I've tracked down the issue with our layout CSS overrides not having the proper precedence. In core's `core/misc/ajax.js`, styles loaded via ajax are prepended to the head of the DOM. This effectively means that **any** CSS loaded via ajax has the **minimum** possible priority, which is pretty problematic as it means all order dependencies are ignored when loading ajax styles. There is an [issue for this here](https://www.drupal.org/project/drupal/issues/1461322)

What's the best approach for a stopgap fix for this? There is a [patch here](https://www.drupal.org/files/issues/2020-02-12/1461322-20.patch) which seems to work. However, it's not a complete solution since it takes the ajax styles and changes the prepend to an append. This is more correct than the current behavior, but the full solution would be that the ajax css gets inserted at weight-appropriate points.

Some options would be that we could use that patch until a more robust solution is available, or we could just change some of our selectors to have more priority so that the low load order doesn't matter as much (though this would potentially mean more convoluted CSS). Any thoughts?

With the patch:

image

**Comment by joeparsons** [comment]
@tadean I'll take a look at that issue but I think I'd be OK with using that patch as a workaround for now.

**Comment by joeparsons** [comment]
As we discussed offline, it might be worth checking to see if we could use an implementation of https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Form%21form.api.php/function/hook_ajax_render_alter/8.9.x to fix when our specific CSS files are loaded before committing to maintaining the core patch.

**Comment by tadean** [comment]
I like that idea as well. If there's enough context to tell it just to append only our styles that seems preferable to maintaining a core patch.

**Comment by tadean** [comment]
Added the mentioned ajax css load order issue patch for testing/evaluation. It would be preferable to employ an implementation that only targets our specific css files, but [hook_ajax_render_alter()](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Form%21form.api.php/function/hook_ajax_render_alter/8.9.x) doesn't seem to have sufficient context to determine that as far as I can tell. I'm checking for possibilities a bit earlier in the rendering process that might be able to determine the context of which css libraries are being included in the ajax request.

This patch should be replaced by an implementation that only affects our CSS libraries, if practical.

**Comment by tadean** [comment]
Added a functional test that tests if our modified workflow of adding a custom block operates correctly, e.g. skips the system block listing.

**Comment by tadean** [comment]
Set it up so that the altered layout workflow (skipping directly to the custom blocks screen choice) only happens for flexible pages, and refactored block css into individual block libraries using the new block-library-detection functionality that was introduced earlier in this PR.

**Comment by joeparsons** [comment]
Followup documentation tasks:
- Document how we add new block types (in a README file in the az_blocks module?)
- Document which customizations affect Flexible pages only vs all Layout Builder things.

**Comment by joeparsons** [comment]
Followup ticket:
- Make accordion Javascript ready for es6 and babel transpilation.

**Pull request merged**